### ISSUE TITLE
DateInput allowEmpty={true} set value to "" instead of null

### ISSUE THREAD

**Issue by kesla** [marmelab/react-admin#3171]
**What you were expecting:**
Creating/editing a form with a DateInput that allows empty sets the value to '' if it has been set before

**What happened instead:**
Value is set to "" instead of `null`

**Steps to reproduce:**
1. Create a DateInput-field that allows the value to be empty
2. Set the field to a date
3. Remove the data
4. Save
5. The field is now "" instead of null

**Related code:**

```js
<DateInput source="foo" allowEmpty />
```

**Other information:**

**Environment**

* React-admin version:
* Last version that did not exhibit the issue (if applicable):
* React version:
* Browser:
* Stack trace (in case of a JS error):

**Comment by kesla** [comment]
I'd be willing to look at this, should be simple to figure out what's wrong :)

**Comment by fargito** [comment]
@kesla did you have time to investigate ?

I was trying to look into it but I think that the `value` was already `""` when the page loaded. Can you confirm that stil is still happening ? 

Also maybe my problem is that I was looking at the wrong `value`. I was looking at the `value` argument of the html `<input />` element, is that what you meant ?

**Comment by Kmaschta** [comment]
Tested and confirmed in latest version https://codesandbox.io/s/simple-rs42q

**Comment by fzaninotto** [comment]
I don't understand this bug. `allowEmpty` is a `ReferenceInput` prop, it should not be used in a `DateInput`. 

Most inputs allow empty value by default - it's up to the developer to add a `required` validator.

**Comment by djhi** [comment]
I don't get the issue here. The empty string is a correct default value for date inputs. If you need to convert it to `null` for your backend, you can either use the `parse` prop from `redux-form` (see our [documentation about it](https://marmelab.com/react-admin/Inputs.html#transforming-input-value-tofrom-record)), or convert it inside your dataProvider.

Feel free to comment if you disagree or ask on Stack Overflow if you need help

**Comment by kesla** [comment]
@djhi We figured it out in our setup so this is not an issue for us right now.

But yeah, I disagree that this is not an issue.

If I add a DateInput and doesn't change it and save/update, the value `null` will be sent to the server. If I changed the value and then remove it the value `""` is sent to the server - even though the state is the same (the value isn't set). That's not what I expected.

But hey, since we've figure this out on our end and you don't seem to think it's an issue - I'm not gonna bother creating a PR for this Just wanted to clarify why I posted the bug report to begin with, so there's no confusion.
